python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,438 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "QuantumGraphExecutionReport",
+    "TaskExecutionReport",
+    "DatasetTypeExecutionReport",
+    "lookup_quantum_data_id",
+)
+
+import dataclasses
+import itertools
+import logging
+import uuid
+from collections.abc import Iterable, Mapping
+from typing import Any
+
+import networkx
+import yaml
+from lsst.daf.butler import Butler, DataCoordinate, DatasetRef
+from lsst.resources import ResourcePathExpression
+
+from .graph import QuantumGraph, QuantumNode
+from .pipeline import PipelineDatasetTypes
+
+
+@dataclasses.dataclass
+class DatasetTypeExecutionReport:
+    """A report on the number of produced datasets as well as the status of
+    missing datasets based on metadata.
+
+    A DatasetTypeExecutionReport is created for each DatasetType in a
+    TaskExecutionReport.
+
+    See Also
+    --------
+    TaskExecutionReport
+    QuantumGraphExecutionReport
+    """
+
+    missing_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced because their quanta failed directly in this
+    run (`set`).
+    """
+    missing_not_produced: dict[DatasetRef, bool] = dataclasses.field(default_factory=dict)
+    """Missing datasets which were not produced due either missing inputs or a
+    failure in finding inputs (`dict`).
+        bool: were predicted inputs produced?
+    """
+    missing_upstream_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced due to an upstream failure (`set`).
+    """
+    n_produced: int = 0
+    """Count of datasets produced (`int`).
+    """
+
+    def to_summary_dict(self) -> dict[str, Any]:
+        """Summarize the DatasetTypeExecutionReport in a dictionary.
+
+        Returns
+        -------
+        A count of the DatasetTypes with each outcome; the number of produced,
```

Comment:
```suggestion
        A count of the datasets with each outcome; the number of produced,
```
This is a summary for a `DatasetType`, but the things we're counting are the actual datasets.  Also a couple of lines below.  And you can put backticks around the attribute names to make them links.